src/main/java/org/jenkinsci/plugins/durabletask/BourneShellScript.java

```diff
@@ -207,6 +188,68 @@ public String getScript() {
         return c;
     }
 
+    private String binaryLauncherCmd(ShellController c, FilePath ws, OsType os, String interpreter, String scriptPath, String cookieValue, String cookieVariable) throws IOException, InterruptedException {
+        String logFile = c.getLogFile(ws).toURI().getRawPath();
+        String resultFile = c.getResultFile(ws).toURI().getRawPath();
+        FilePath controlDir = c.controlDir(ws);
+        String outputFile = c.getOutputFile(ws).toURI().getRawPath();
+
+        String launcher_binary = LAUNCHER_PREFIX + os.toString();
+
+        InputStream launcherStream = this.getClass().getResourceAsStream(launcher_binary);
+        FilePath launcherAgent = controlDir.child(launcher_binary);
+        launcherAgent.copyFrom(launcherStream);
+        launcherAgent.chmod(0755);
+
+        String args = String.format("%s %s %s '%s %s'",
+                        controlDir.toURI().getRawPath(),
+                        resultFile,
+                        logFile,
+                        interpreter, scriptPath);
+        if (capturingOutput) {
+            args = args + " " + outputFile;
+        }
+        String cmd = String.format("jsc=%s; %s=$jsc %s %s; ",
+                        cookieValue,
+                        cookieVariable,
+                        launcherAgent.getRemote(),
+                        args);
+
+        return cmd;
+    }
+
+    private String scriptLauncherCmd(ShellController c, FilePath ws, String interpreter, String scriptPath, String cookieValue, String cookieVariable) throws IOException, InterruptedException {
+        String cmd;
+        FilePath logFile = c.getLogFile(ws);
+        FilePath resultFile = c.getResultFile(ws);
+        FilePath controlDir = c.controlDir(ws);
+        if (capturingOutput) {
+            cmd = String.format("pid=$$; { while [ \\( -d /proc/$pid -o \\! -d /proc/$$ \\) -a -d '%s' -a \\! -f '%s' ]; do touch '%s'; sleep 3; done } & jsc=%s; %s=$jsc %s '%s' > '%s' 2> '%s'; echo $? > '%s.tmp'; mv '%s.tmp' '%s'; wait",
+                    controlDir,
+                    resultFile,
+                    logFile,
+                    cookieValue,
+                    cookieVariable,
+                    interpreter,
+                    scriptPath,
+                    c.getOutputFile(ws),
+                    logFile,
+                    resultFile, resultFile, resultFile);
+        } else {
+            cmd = String.format("pid=$$; { while [ \\( -d /proc/$pid -o \\! -d /proc/$$ \\) -a -d '%s' -a \\! -f '%s' ]; do touch '%s'; sleep 3; done } & jsc=%s; %s=$jsc %s '%s' > '%s' 2>&1; echo $? > '%s.tmp'; mv '%s.tmp' '%s'; wait",
```

Comment:
As I think I mentioned previously, we expect almost all users to be running the binary launcher so we can strip out some of the hard-to-test exotica in this case. Probably drop `nohup` (already did that?); and the `/proc` heartbeat trick (`/proc` will only work on Linux anyway!).